Tests/RevenueCatUITests/Templates/ExternalPurchaseAndRestoreTests.swift

```diff
@@ -0,0 +1,393 @@
+            return (userCancelled: false, error: TestError.error)
+        } performRestore: {
+            callbackOrder.append("performRestore")
+            return (success: true, error: nil)
+        }
+
+        try PaywallView(
+            offering: Self.offering.withLocalImages,
+            customerInfo: TestData.customerInfo,
+            introEligibility: .producing(eligibility: .eligible),
+            purchaseHandler: purchasHandler
+        )
+        .onPurchaseStarted { _ in
+            callbackOrder.append("onPurchaseStarted")
+        }
+        .onPurchaseCompleted { _ in
+            callbackOrder.append("onPurchaseCompleted")
+        }
+        .onPurchaseCancelled {
+            callbackOrder.append("onPurchaseCancelled")
+        }
+        .onPurchaseFailure { error in
+            callbackOrder.append("onPurchaseFailure")
+            passedError = error
+        }
+        .onRestoreStarted({
+            callbackOrder.append("onRestoreStarted")
+        })
+        .onRestoreCompleted({ _ in
+            callbackOrder.append("onRestoreCompleted")
+        })
+        .onRestoreFailure({ _ in
+            callbackOrder.append("onRestoreFailure")
+        })
+        .addToHierarchy()
+
+        Task {
+            _ = try await purchasHandler.purchase(package: Self.package)
+        }
+
+        expect(callbackOrder).toEventually(equal(["onPurchaseStarted", "performPurchase", "onPurchaseFailure"]))
+        expect(passedError).toEventually(matchError(TestError.error))
+    }
+
+    func testHandleExternalRestore() throws {
+        var completed = false
+        var customerInfo: CustomerInfo?
+        var callbackOrder = [String]()
+
+        let purchasHandler = Self.externalPurchaseHandler { _ in
+            return (userCancelled: true, error: nil)
+        } performRestore: {
+            callbackOrder.append("performRestore")
+            return (success: true, error: nil)
+        }
+
+        try PaywallView(
+            offering: Self.offering.withLocalImages,
+            customerInfo: TestData.customerInfo,
+            introEligibility: .producing(eligibility: .eligible),
+            purchaseHandler: purchasHandler
+        )
+        .onPurchaseStarted { _ in
+            callbackOrder.append("onPurchaseStarted")
+        }
+        .onPurchaseCompleted { _ in
+            callbackOrder.append("onPurchaseCompleted")
+        }
+        .onPurchaseCancelled {
+            callbackOrder.append("onPurchaseCancelled")
+        }
+        .onPurchaseFailure { _ in
+            callbackOrder.append("onPurchaseFailure")
+        }
+        .onRestoreStarted({
+            callbackOrder.append("onRestoreStarted")
+        })
+        .onRestoreCompleted({ info in
+            callbackOrder.append("onRestoreCompleted")
+            customerInfo = info
+        })
+        .onRestoreFailure({ _ in
+            callbackOrder.append("onRestoreFailure")
+        })
+        .addToHierarchy()
```

Comment:
I don't mind duplicate code in tests usually but it feels like this is an easy extract?

Reply:
Ugh I know, I almost mentioned this in the PR writeup. I tried to extract it a few ways but it's not easy, or at least I didn't find a way, I was being blocked because `callbackOrder.append("")` is a mutating function and these modifiers are all escaping closures and it is not allowed to do this. I tried to work around it but came up blank and decided for testing code it wasn't worth continuing to work on. IF there is a good way to extract it though I'd love to use it!!